lab_4/main.py

```diff
@@ -3,31 +3,92 @@
 """
 
 from ngrams.ngram_trie import NGramTrie
-
+import re
 
 def tokenize_by_sentence(text: str) -> tuple:
-    pass
+    if not isinstance(text, str):
+        raise ValueError
+    if not text:
+        return ()
+
+    text_splited = re.split('[!?.]', text)
+    new_text = []
+
+    for sentence in text_splited:
+        if len(sentence.strip()):
+            list_tokens = re.sub('[^a-z \n]', '', sentence.lower()).split()
+            if len(list_tokens):
+                list_tokens += ['<END>']
+                new_text.append(list_tokens)
+
+    output = []
+
+    for sentence in new_text:
+        for element in sentence:
+            output.append(element)
+
+    return tuple(output)
 
 
 class WordStorage:
     def __init__(self):
-        pass
+        self.storage = {}
+        self.id = 1
 
     def _put_word(self, word: str):
-        pass
+
+        if not isinstance(word, str):
+            raise ValueError
+
+        if word not in self.storage:
+            self.storage[word] = self.id
+            self.id += 1
+
+        return self.storage[word]
 
     def get_id(self, word: str) -> int:
-        pass
+
+        if not isinstance(word, str):
+            raise ValueError
+
+        for word_another in self.storage:
```

Comment:
There is a much faster way of doing so. Read about accessing the keys/values of a dictionary.